src/artifact/parser.cpp

```diff
@@ -186,26 +186,35 @@ ExpectedArtifact Parse(io::Reader &reader, config::ParserConfig config) {
 		return artifact;
 	}
 
-	log::Trace("Parsing the payload");
-	tok = lexer.Next();
-	if (tok.type != token::Type::Payload) {
-		return expected::unexpected(parser_error::MakeError(
-			parser_error::Code::ParseError,
-			"Got unexpected token " + tok.TypeToString() + " expected 'data/0000.tar"));
-	}
-
 	return artifact;
 };
 
 
 ExpectedPayload Artifact::Next() {
-	// Currently only one payload supported
-	if (payload_index_ != 0) {
-		return expected::unexpected(parser_error::MakeError(
-			parser_error::Code::EOFError, "Reached the end of the Artifact"));
+	token::Token tok = lexer_.Next();
```

Comment:
Nice fixes!

I would suggest something like this though:

```
ExpectedPayload Artifact::Next() {
	token::Token tok = lexer_.Next();
	if (payload_index_ != 0) {
		// Currently only one payload supported
		switch (tok.type) {
		case token::Type::EOFToken:
			return expected::unexpected(parser_error::MakeError(
				parser_error::Code::EOFError, "Reached the end of the Artifact"));
		case token::Type::Payload:
			return expected::unexpected(error::Error(
				make_error_condition(errc::not_supported), "Only one artifact payload supported"));
		default:
			return expected::unexpected(parser_error::MakeError(
				parser_error::Code::ParseError, "Unexpected token: " + tok.TypeToString()));
		};
	}
	if (tok.type != token::Type::Payload) {
		return expected::unexpected(parser_error::MakeError(
			parser_error::Code::ParseError,
			"Got unexpected token " + tok.TypeToString() + " expected 'data/0000.tar"));
	}
	log::Trace("Parsing the payload");
	payload_index_++;
	return payload::Payload(*(this->lexer_.current.value), manifest);
}
```

Same thing, but structured a little differently

Reply:
Agreed, this is clearer.